### ISSUE TITLE
Job System Issues

### ISSUE THREAD

**Issue by halgari** [Nexus-Mods/NexusMods.App#1942]
The Job system is somewhat cumbersome and verbose to work with. Here's a short rundown of some of the issues I ran across trying to implement more complex systems (such as collections downloads) on top of the job system.

- [ ] Jobs split logic from inputs, but combine parameters and state. This means that at a minimum if you want to make a new job you have to write two classes. The "worker" and the state. This becomes cumbersome to maintain, resulting in an insentive not to make new types of jobs. 
- [ ] Jobs don't compose well. Each job that calls another job must somehow figure out its own state and then figure out if a sub job also is in-progress
- [ ] The Pause/Resume/Cancel semantics make the code *extremely* complex, when most jobs will complete so fast they'll never need to be canceled.
- [ ] Jobs are untyped. There's now way to know if a given job returns an `int` a `float` or a `Foo`, to figure out the return type you have to dig into the worker and figure out what the final branch of the code will return. Sometimes Jobs call other jobs and use `var` making this all the more hard to understand
- [ ] Persistent jobs are infectious. If a job wants to download something from a HTTP source, it must be a persistent job, because the HTTP downloader is also persistent. 
- [ ] There is a *lot* of ceremony in the code. Each Job and Worker takes several parameters that do nothing but tie data together (monitors, progress, etc.) and all of this must be implemented for everyone who wants to use the job. 
- [ ] We still don't have rate limiting in any of these job systems, and if we did that would increase the complexity even more. We can't queue up 1000 download jobs, because the worker would just invoke all the jobs in parallel. 
- [ ] What used to be a simple method call, now takes 4-10 lines of code. Since to call any job we have to do the following:
  - [ ] Allocate a Job (1 line)
  - [ ] Start a job (1 line)
  - [ ] Wait for the job to finish (1 line)
  - [ ] Make sure the job completed, and get a untyped result object (via a try/get so that's 2 more lines of code)
  - [ ] Guess on the return type, hope it never changes, and use `TryGet` to get the result type.  (2 lines)
  - [ ] 7 lines of code for what should be 1 line. 
- [ ] I'm pretty convinced the Persistent job is broken when combined with nested tasks. Most of the nested code doesn't properly enforce the infectious nature of persistent jobs (they are transient jobs that use persistent jobs), and the ones that do, don't properly know when to jump back into the internals of a sub job. 

  
# Design Ideas
I'm not sure what we want here, but I don't think this is it. It's interesting that we've somewhat reimplemented `async/await`. Perhaps there's some way to leverage that. 

It might be that we can drastically reduce this complexity by assuming that jobs will be emphemeral, and we can implement more complex logic for persistent jobs. For example having explicit "save/load" states on persistent jobs. Or really, just ignore the persistent bit for now. The *only* thing in our system that is persistent is the HTTP downloader and we can store inprogress state on disk, and use that to early complete any jobs that are in progress. Something like hashing the URL or the source parameters and then implicitly resuming based on this data. 

Another idea here is to make Pause/Cancel/Stop opt-in. Really the only thing that needs most of these is the HTTP jobs. Most other jobs would be a straight up cancel. If we removed that as a hard requirement we could make most jobs a simple async method call. 

# Supporting Examples

In this small example, a good 50% of the code is constructing jobs, just so we can execute them and wait for them. We've turned simple function calls into many lines of boilerplate. This is roughly 25 lines of code, that could probably be 4 lines written conventionally. In addition any code that wants to use this code must now include this "allocate job, start job, end job, try get result" combination. It's a 4x bloat in lines of code for something that could be `var result = worker.Execute(param1)`

```csharp
    protected override async Task<JobResult> ExecuteAsync(AddDownloadJob job, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (!job.DownloadJobResult.HasValue)
        {
            await job.DownloadJob.StartAsync(cancellationToken: cancellationToken);
            var result = await job.DownloadJob.WaitToFinishAsync(cancellationToken: cancellationToken);
            job.DownloadJobResult = result.RequireData<AbsolutePath>();
        }

        cancellationToken.ThrowIfCancellationRequested();
        using var tx = _connection.BeginTransaction();

        await using var addLibraryFileJob = new AddLibraryFileJob(group: job, monitor: _jobMonitor, worker: _serviceProvider.GetRequiredService<AddLibraryFileJobWorker>())
        {
            Transaction = tx,
            FilePath = job.DownloadJobResult.Value,
            DoBackup = true,
            DoCommit = false,
        };

        await addLibraryFileJob.StartAsync(cancellationToken: cancellationToken);
        var addLibraryFileJobResult = await addLibraryFileJob.WaitToFinishAsync(cancellationToken: cancellationToken);

        var libraryFile = addLibraryFileJobResult.RequireData<LibraryFile.New>();
        await job.DownloadJob.AddMetadata(tx, libraryFile);

        var transactionResult = await tx.Commit();
        return JobResult.CreateCompleted(transactionResult.Remap(libraryFile));
    }
```

This is the guts of the "Add Library" code.  If we go through and delete all the bookkeeping code, we can drastically reduce the complexity. With all the "Save Data" / "Check if Cancelled" / "Load Data" boilerplate it's hard to even figure out what the code is doing. This is a state machine and we're managing it all by hand. 

```csharp
    protected override async Task<JobResult> ExecuteAsync(AddLibraryFileJob job, CancellationToken cancellationToken)
    {
        var absolutePath = job.FilePath;
        if (!absolutePath.FileExists)
        {
            if (absolutePath.DirectoryExists())
            {
                return JobResult.CreateFailed($"File at `{absolutePath}` can't be added to the library because it's a directory");
            }

            return JobResult.CreateFailed($"File at `{absolutePath}` can't be added to the library because it doesn't exist");
        }

        cancellationToken.ThrowIfCancellationRequested();
        if (!job.IsArchive.HasValue)
        {
            job.IsArchive = await CheckIfArchiveAsync(job.FilePath);
        }

        cancellationToken.ThrowIfCancellationRequested();
        if (!job.HashJobResult.HasValue)
        {
            job.HashJobResult = await HashAsync(job.FilePath, cancellationToken);
        }

        cancellationToken.ThrowIfCancellationRequested();
        if (!job.EntityId.HasValue)
        {
            job.EntityId = job.Transaction.TempId();
        }

        cancellationToken.ThrowIfCancellationRequested();
        if (!job.LibraryFile.HasValue)
        {
            job.LibraryFile = CreateLibraryFile(
                job.Transaction,
                job.EntityId.Value,
                job.FilePath,
                hash: job.HashJobResult.Value.RequireData<Hash>()
            );
        }

        if (job.IsArchive.Value)
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (!job.ExtractionDirectory.HasValue)
            {
                job.ExtractionDirectory = _temporaryFileManager.CreateFolder();
            }

            cancellationToken.ThrowIfCancellationRequested();
            if (!job.ExtractedFiles.HasValue)
            {
                var extractedFiles = await ExtractArchiveAsync(
                    job,
                    job.FilePath,
                    job.ExtractionDirectory.Value,
                    cancellationToken
                );

                if (extractedFiles.Length == 0)
                {
                    _logger.LogWarning("File `{Path}` was assumed to be extractable but no files were extracted, it will not be added as an archive", job.FilePath);
                    job.IsArchive = false;
                }
                else
                {
                    job.ExtractedFiles = extractedFiles;
                }
            }
        }

        if (job.IsArchive.Value)
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (!job.LibraryArchive.HasValue)
            {
                job.LibraryArchive = new LibraryArchive.New(job.Transaction, job.EntityId.Value)
                {
                    LibraryFile = job.LibraryFile.Value,
                    IsArchive = true,
                };
            }

            cancellationToken.ThrowIfCancellationRequested();
            if (!job.AddExtractedFileJobResults.HasValue)
            {
                var extractedFiles = job.ExtractedFiles.Value;
                var results = new ValueTuple<JobResult, IFileEntry>[extractedFiles.Length];

                await Parallel.ForAsync(fromInclusive: 0, toExclusive: extractedFiles.Length, cancellationToken, async (i, innerCancellationToken) =>
                {
                    var fileEntry = extractedFiles[i];

                    var worker = _serviceProvider.GetRequiredService<AddLibraryFileJobWorker>();
                    await using var childJob = new AddLibraryFileJob(job, worker)
                    {
                        Transaction = job.Transaction,
                        FilePath = fileEntry.Path,
                        DoCommit = false,
                        DoBackup = false,
                    };

                    await worker.StartAsync(childJob, cancellationToken: innerCancellationToken);
                    var result = await childJob.WaitToFinishAsync(cancellationToken: innerCancellationToken);

                    results[i] = (result, fileEntry);
                });

                job.AddExtractedFileJobResults = results;
            }

            cancellationToken.ThrowIfCancellationRequested();
            if (!job.HasBackup.HasValue)
            {
                var filesToBackup = job.AddExtractedFileJobResults.Value
                    .Select(tuple =>
                    {
                        var (jobResult, fileEntry) = tuple;
                        var data = jobResult.RequireData<LibraryFile.New>();

                        return new ArchivedFileEntry
                        {
                            Hash = data.Hash,
                            Size = data.Size,
                            StreamFactory = new NativeFileStreamFactory(fileEntry.Path),
                        };
                    })
                    .ToArray();

                await _fileStore.BackupFiles(filesToBackup, token: cancellationToken);
                job.HasBackup = true;
            }

            cancellationToken.ThrowIfCancellationRequested();
            foreach (var tuple in job.AddExtractedFileJobResults.Value)
            {
                var (jobResult, fileEntry) = tuple;
                var libraryFile = jobResult.RequireData<LibraryFile.New>();
                var path = fileEntry.Path.RelativeTo(job.ExtractionDirectory.Value.Path);

                _ = new LibraryArchiveFileEntry.New(job.Transaction, libraryFile.Id)
                {
                    Path = path,
                    LibraryFile = libraryFile,
                    ParentId = job.LibraryArchive.Value,
                };
            }
        }
        else
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (job is { DoBackup: true, HasBackup.HasValue: false })
            {
                var archivedFileEntry = new ArchivedFileEntry
                {
                    Hash = job.HashJobResult.Value.RequireData<Hash>(),
                    Size = job.FilePath.FileInfo.Size,
                    StreamFactory = new NativeFileStreamFactory(job.FilePath),
                };

                await _fileStore.BackupFiles([archivedFileEntry], token: cancellationToken);
                job.HasBackup = true;
            }
        }

        cancellationToken.ThrowIfCancellationRequested();
        if (job.DoCommit)
        {
            var transactionResult = await job.Transaction.Commit();
            return JobResult.CreateCompleted(transactionResult.Remap(job.LibraryFile.Value));
        }

        return JobResult.CreateCompleted(job.LibraryFile.Value);
    }
```

**Comment by erri120** [comment]
> Jobs split logic from inputs, but combine parameters and state. This means that at a minimum if you want to make a new job you have to write two classes. The "worker" and the state. This becomes cumbersome to maintain, resulting in an insentive not to make new types of jobs.

There is some amount of boilerplate required. It's not as bad as all the things we need to do to add a new View and VM. You can also make an "anonymous worker" using the `JobWorker.Create` method. That saves on creating a worker, but you still need the job type.

> The Pause/Resume/Cancel semantics make the code _extremely_ complex, when most jobs will complete so fast they'll never need to be canceled.

It's true that the actual code behind executing a job is non-trivial, but I wouldn't call it complex. The actual code in the worker should be super straightforward, though. As for cancellation, we can't make an assumption on how fast these jobs complete. It's true that some jobs will finish quickly, hashing a 1KB file will finish before the UI even updated, but hashing a 10GB file on an HDD will take a while.

> Persistent jobs are infectious. If a job wants to download something from a HTTP source, it must be a persistent job, because the HTTP downloader is also persistent.

I don't think they are infectious. If `Foo` is a job that downloads something, and `Foo` isn't persistent, then the download job might be picked up again on startup, but it wouldn't really do anything. I don't think our system will implode with that. It's definitely weird, but it very much depends on what the job is doing. The entire "persistence" thing was made only with downloads in mind, and both me and @Al12rs couldn't think of any other stuff we might want to persist.

> What used to be a simple method call, now takes 4-10 lines of code.

As mentioned before, we can make a single call to handle all that.

> I'm not sure what we want here, but I don't think this is it. It's interesting that we've somewhat reimplemented `async/await`. Perhaps there's some way to leverage that.

Yup, it's basically async-await where the jobs are tasks that can be restarted. We even hijack `CancellationToken` for both cancel and pause signaling.

> Another idea here is to make Pause/Cancel/Stop opt-in. Really the only thing that needs most of these is the HTTP jobs. Most other jobs would be a straight up cancel. If we removed that as a hard requirement we could make most jobs a simple async method call.

Only pause should be opt-in, every task/job should be cancellable. You'd have two variants:

1) normal job, can start and cancel, can't be paused
2) "pausable" job, can start, cancel, pause, and resume

That'd be nice to have, similar to having the output on the job type itself, it's just going to require a more complex abstraction. My initial job design was very complex with different interfaces for jobs with output types, read-only jobs, mutable jobs, and whatnot, and it was a complete mess. We can try again now, since we have a better understanding of what we want from the system.

**Comment by halgari** [comment]
Having slept on it a bit and thinking through the above comments the following may get me what I want: 

1) move job creation into the workers, this way I don't need to pass around `_serviceProvider` and `_jobMonitor` if all I want is to make a new job, I can do `downloader.CreateJob`, and pass in the parameters that make sense to my code.
2) Make a new `PausableToken` that wraps the semantics of `CancellationToken` this way we can pause any job whenever tehy do `await token.Handle()`. It'll throw if cancellation is requested, or just pause waiting for the await if the execution is paused. 
3) Since we now have pause and cancel handle natively we can get rid of the `Job` types and use async methods. If we want to monitor/track jobs we can do so via the `token` class. 
4) Move persistence into the job worker or something, I don't think this is hard to handle but since only one job needs it at the moment we can probably be fine with making the code for it fairly special case.

**Comment by halgari** [comment]
I'm going to rewrite this code, as an example and see what this would look like, one sec...

```csharp
using Microsoft.Extensions.DependencyInjection;
using NexusMods.Abstractions.Jobs;
using NexusMods.Abstractions.Library.Models;
using NexusMods.MnemonicDB.Abstractions;
using NexusMods.Paths;

namespace NexusMods.Library;

internal class AddDownloadJobWorker : AJobWorker<AddDownloadJob>
{
    private readonly IServiceProvider _serviceProvider;
    private readonly IJobMonitor _jobMonitor;
    private readonly IConnection _connection;

    public AddDownloadJobWorker(IServiceProvider serviceProvider)
    {
        _serviceProvider = serviceProvider;
        _jobMonitor = serviceProvider.GetRequiredService<IJobMonitor>();
        _connection = serviceProvider.GetRequiredService<IConnection>();
    }

    protected override async Task<JobResult> ExecuteAsync(AddDownloadJob job, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (!job.DownloadJobResult.HasValue)
        {
            await job.DownloadJob.StartAsync(cancellationToken: cancellationToken);
            var result = await job.DownloadJob.WaitToFinishAsync(cancellationToken: cancellationToken);
            job.DownloadJobResult = result.RequireData<AbsolutePath>();
        }

        cancellationToken.ThrowIfCancellationRequested();
        using var tx = _connection.BeginTransaction();

        await using var addLibraryFileJob = new AddLibraryFileJob(group: job, monitor: _jobMonitor, worker: _serviceProvider.GetRequiredService<AddLibraryFileJobWorker>())
        {
            Transaction = tx,
            FilePath = job.DownloadJobResult.Value,
            DoBackup = true,
            DoCommit = false,
        };

        await addLibraryFileJob.StartAsync(cancellationToken: cancellationToken);
        var addLibraryFileJobResult = await addLibraryFileJob.WaitToFinishAsync(cancellationToken: cancellationToken);

        var libraryFile = addLibraryFileJobResult.RequireData<LibraryFile.New>();
        await job.DownloadJob.AddMetadata(tx, libraryFile);

        var transactionResult = await tx.Commit();
        return JobResult.CreateCompleted(transactionResult.Remap(libraryFile));
    }
}

```

**Comment by halgari** [comment]
I think we can get that code down to something like this, perhaps even shorter depending on how much more we want to code golf.

```csharp
internal class AddDownloadJobWorker : AJobWorker<AddDownloadJob>
{
    private readonly IServiceProvider _serviceProvider;
    private readonly IConnection _connection;

    public AddDownloadJobWorker(IServiceProvider serviceProvider)
    {
        _serviceProvider = serviceProvider;
        _connection = serviceProvider.GetRequiredService<IConnection>();
    }

    protected override async Task<JobResult> ExecuteAsync(AddDownloadJob job, JobToken token)
    {
        await token;
        var downloadPath = await job;
        
        await token;
        using var tx = _connection.BeginTransaction();
        var libraryItem = _serviceProvider.GetWorker<AddLibraryFileJobWorker>().AddLibraryFileJob(tx, job, downloadPath);
        await token;
        
        await AddMetadata(tx, libraryItem);
        await token;
        
        var transactionResult = await tx.Commit();
        return transactionResult.Remap(libraryItem);
    }
}
```

**Comment by Al12rs** [comment]
### Meeting Notes:

#### Job System requirements:
- [ ] Rate limiting job operation to prevent system saturation
	- [ ] Recursive jobs can have issues with rate limiting system
- [ ] Knowing what jobs are currently in flight (job monitor or download job manager)
- [ ] Some jobs need to report progress
	- [ ] Some jobs don't have progress to report except that it is progressing
- [ ] All jobs should support cancellation
- [ ] Some jobs need to support pausing/resuming (downloads)
	- [ ] Do we have other examples?
- [ ] The downloads jobs need to support app session persistence
	- [ ] Will we have other jobs that will need app session persistence
- [ ] Composability of jobs:
	- [ ] Being able to tie together multiple jobs into a single one in a concise manner
- [ ] Statically typed job results

**Comment by halgari** [comment]
Taken care of as part of #1983, will file additional issues as needed
